### PR TITLE
Add -fno-builtin to dlmalloc / math library build

### PR DESCRIPTION
**Pull request by aheejin**
They should be built with `-fno-builtin` because otherwise LLVM can optimize their callsites in various ways assuming they are functions from the standard library.

This fixes [the test_regex failure in the current wasm waterfall](https://logs.chromium.org/v/?s=chromium%2Fbb%2Fclient.wasm.llvm%2Flinux%2F29568%2F%2B%2Frecipes%2Fsteps%2FExecute_emscripten_testsuite__emwasm_%2F0%2Fstdout).

Here I added the flag to dlmalloc and math library. Should we add `-fno-builtin` in other library builds as well?

### CODE REVIEW

**Comment by kripken** [comment]
This makes sense, but I'm confused by the error on the waterfall - I'd expect it to have a linking error or such if it's trying to use a function that isn't defined, but it looks like the error is incorrect execution?

Also, I'm confused by dlmalloc - which builtin is it optimizing there?

**Comment by aheejin** [comment]
Library functions in `dlmalloc.c`, such as `dlmalloc` or `dlcalloc`, are converted to `malloc` and `calloc` by the preprocessor before being compiled, so there is no linking errors.
https://github.com/kripken/emscripten/blob/4c1c351d03c451873459444444a9fd964943475f/system/lib/dlmalloc.c#L850-L875

The problematic callsite is here, where `calloc` calls `malloc`.
https://github.com/kripken/emscripten/blob/4c1c351d03c451873459444444a9fd964943475f/system/lib/dlmalloc.c#L4846-L4849

`mem2chunk(mem)` is a macro that effectively means `mem - 8`. So, this accesses memory before the malloc'ed pointer, which is UB in the standard malloc. LLVM's GVN pass identifies this kind of callsites and converts them to `undef` unless it is compiled with `-fno-builtin`. So the whole `if (mem != 0 && calloc_must_clear(mem2chunk(mem))) ...` part gets optimized away and it ends up not running `memset` for both cases.

---

More detail just in case you are interested:

Before GVN:
```llvm
...

if.end3:
  %req.1 = phi i64 [ 0, %entry ], [ %mul, %if.then ], [ %spec.select, %land.lhs.true ]
  %call = tail call noalias i8* @malloc(i64 %req.1) #9
  %cmp4 = icmp eq i8* %call, null
  br i1 %cmp4, label %if.end9, label %land.lhs.true5

land.lhs.true5:
  %head = getelementptr inbounds i8, i8* %call, i64 -8 ; Out of bounds access, which is UB
  %0 = bitcast i8* %head to i64*
  %1 = load i64, i64* %0, align 8      ; THIS LOAD WILL BE CONVERTED TO UNDEF !!!
  %and6 = and i64 %1, 3
  %cmp7 = icmp eq i64 %and6, 0
  br i1 %cmp7, label %if.end9, label %if.then8

if.then8:
  tail call void @llvm.memset.p0i8.i64(i8* nonnull align 1 %call, i8 0, i64 %req.1, i1 false)
  br label %if.end9

if.end9:
  ret i8* %call
```

After GVN:
```llvm
...

if.end3:
  %req.1 = phi i64 [ 0, %entry ], [ %mul, %if.then ], [ %spec.select, %land.lhs.true ]
  %call = tail call noalias i8* @malloc(i64 %req.1) #9
  %cmp4 = icmp eq i8* %call, null
  br i1 %cmp4, label %if.end9, label %land.lhs.true5

land.lhs.true5:
  %head = getelementptr inbounds i8, i8* %call, i64 -8
  %0 = bitcast i8* %head to i64*
  ; Because the 'load' and 'and' insts were optimized away, now this branch is meaningless
  br i1 true, label %if.end9, label %if.then8

if.then8:
  tail call void @llvm.memset.p0i8.i64(i8* nonnull align 1 %call, i8 0, i64 %req.1, i1 false)
  br label %if.end9

if.end9:
  ret i8* %call
```

At the end, this becomes:
```llvm
...

if.end3:
  %req.1 = phi i64 [ 0, %entry ], [ %mul, %if.then ], [ %spec.select, %land.lhs.true ]
  %call = tail call noalias i8* @malloc(i64 %req.1) #9
  ret i8* %call
```
which essentially wipes out the `memset` part altogether.

---

It had been OK so far, because, before llvm-mirror/llvm@c138405, if `malloc`'s definition was in the same compilation unit, it did not optimize malloc's callsites. But that commit somehow removed that condition, which I think is also a correct behavior. Whether `malloc` is within the same compilation unit or not should not necessarily cause any difference because it's entirely possible that we have our custom `malloc` implementation in another file.

__So, long story short, we should disable implicit builtin knowledge on memory allocation functions for `dlmalloc`, because what they are doing is UB for the standard `malloc`.__ 

Also, LLVM has various optimizations targeting builtin math functions. For example, @dschuff said there once had been a similar bug for PNaCl: their custom math library implements `exp` in terms of `pow`, and LLVM has an optimization that turns `pow` with base `e` into `exp`, so not compiling them with `-fno-builtin` caused an infinite recursion.

**Comment by kripken** [comment]
Wow, very interesting! I had no idea gvn did optimizations like that. But it makes total sense now that I get it.

**Comment by kripken** [comment]
Maybe worth adding comments in the source to explain this there?

**Comment by aheejin** [comment]
I added comments, but there are several places that I added `-fno-builtin`, so not sure how many times I should duplicate the comments.

**Comment by kripken** [comment]
Hmm, yeah, maybe just commenting before the first one is enough.

**Comment by kripken** [comment]
Thanks, lgtm.
